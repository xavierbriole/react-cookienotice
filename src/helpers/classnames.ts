export default (...args: Array<any>): string => {
  const classes = []

  for (const arg of args) {
    if (arg) {
      if (typeof arg === 'string') {
        classes.push(arg)
      } else if (Array.isArray(arg)) {
        classes.push(...arg)
      } else {
        for (const key in arg) {
          if (arg[key]) {
            classes.push(key)
          }
        }
      }
    }
  }

  return classes.join(' ')
}
