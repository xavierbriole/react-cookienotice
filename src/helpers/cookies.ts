export const setCookie = (options: CookieOptions): void => {
  const { name, value, expires, secure, httpOnly, sameSite } = options

  const date = new Date()
  date.setTime(date.getTime() + expires * 24 * 60 * 60 * 1000)

  let cookie = `${name}=${value};expires=${date.toUTCString()};path=/;`

  if (secure) cookie += 'Secure;'
  if (httpOnly) cookie += 'HttpOnly;'

  cookie += `SameSite=${sameSite};`

  document.cookie = cookie
}

export const getCookieValue = (name: string): string | null => {
  const value = document.cookie.match(`(^|;) ?${name}=([^;]*)(;|$)`)

  return value ? decodeURIComponent(value[2]) : null
}
