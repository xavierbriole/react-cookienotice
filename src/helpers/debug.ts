export const err = (message: string): void => {
  if (process.env.NODE_ENV === 'development') {
    console.error(`[react-cookienotice] ${message}`)
  }
}
