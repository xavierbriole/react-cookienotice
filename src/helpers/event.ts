interface ReactCookieNoticeEvent {
  type: 'REACT_COOKIENOTICE_EVENT'
  service: string
  enabled: boolean
}

export default class Event {
  static send(event: ReactCookieNoticeEvent): void {
    window.postMessage(event, '*')
  }
}
