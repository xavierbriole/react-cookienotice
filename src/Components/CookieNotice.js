// @flow

import React, { type ComponentType } from 'react'
import clsx from 'clsx'
import styled from 'styled-components'
import AcceptButton from './AcceptButton'
import ReadMoreButton from './ReadMoreButton'
import CookieIcon from './CookieIcon'
import CookieText from './CookieText'
import {
  validateAcceptButtonLabel,
  validateReadMoreButtonLabel,
  validateReadMoreButtonLink,
  validateOpenInNewTab,
  validateCookieTextLabel,
  validateReverseButtons,
  validateBorderRadius,
  validateJustifyContent,
  validateMaxWidth,
} from '../Validator'
import { getCookie, setCookie } from '../Helpers/cookies'

type WrapperProps = {
  justifyContent: 'space-around' | 'space-between',
  maxWidth: number,
  borderRadius: number,
}

const Root = styled.div`
  opacity: 1;
  transition: opacity 0.5s linear;

  &.cookies-allowed {
    opacity: 0;
  }
`

const StickToBottom = styled.div`
  position: fixed;
  left: 0px;
  bottom: 0px;
  width: 100%;
  z-index: 1201;
`

const Wrapper: ComponentType<WrapperProps> = styled.div`
  display: flex;
  justify-content: ${({ justifyContent }) => justifyContent};
  align-items: center;
  max-width: ${({ maxWidth }) => maxWidth}px;
  opacity: 1;
  box-shadow: rgba(0, 0, 0, 0.1) 0px 24px 32px, rgba(0, 0, 0, 0.1) 0px 8px 32px;
  margin: 0 auto 48px;
  background: #ffffff;
  padding: 16px;
  border-radius: ${({ borderRadius }) => borderRadius}px;

  @media (max-width: ${({ maxWidth }) => maxWidth + 48}px) {
    margin-left: 48px;
    margin-right: 48px;
  }

  @media (prefers-color-scheme: dark) {
    background: #3a3a3a;
  }
`

const ButtonsWrapper = styled.div`
  display: flex;
  align-items: center;

  @media (max-width: 768px) {
    flex-direction: column;
  }
`

type Props = {|
  acceptButtonLabel?: string,
  readMoreButtonLabel?: string,
  readMoreButtonLink?: string,
  openInNewTab?: boolean,
  cookieTextLabel?: string,
  reverseButtons?: boolean,
  borderRadius?: number,
  justifyContent?: 'space-around' | 'space-between',
  maxWidth?: number,
|}

type State = {|
  cookiesAllowed: boolean,
|}

export default class CookieNotice extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)

    const cookieValue = getCookie('allow-cookies') === 'true'

    this.state = { cookiesAllowed: cookieValue }
  }

  setCookie() {
    setCookie('allow-cookies', 'true')

    this.setState({ cookiesAllowed: true })
  }

  render() {
    const {
      acceptButtonLabel,
      readMoreButtonLabel,
      readMoreButtonLink,
      openInNewTab,
      cookieTextLabel,
      reverseButtons,
      borderRadius,
      justifyContent,
      maxWidth,
    } = this.props

    const { cookiesAllowed } = this.state

    const buttons = [
      <AcceptButton
        key='accept-button'
        label={validateAcceptButtonLabel(acceptButtonLabel)}
        onButtonClick={() => this.setCookie()}
      />,
      <ReadMoreButton
        key='read-more-button'
        label={validateReadMoreButtonLabel(readMoreButtonLabel)}
        link={validateReadMoreButtonLink(readMoreButtonLink)}
        openInNewTab={validateOpenInNewTab(openInNewTab)}
      />,
    ]

    return (
      <Root
        className={clsx('cookie-notice-root', {
          'cookies-allowed': cookiesAllowed,
        })}
      >
        <StickToBottom className='stick-to-bottom'>
          <Wrapper
            className='wrapper'
            borderRadius={validateBorderRadius(borderRadius)}
            justifyContent={validateJustifyContent(justifyContent)}
            maxWidth={validateMaxWidth(maxWidth)}
          >
            <CookieIcon />
            <CookieText label={validateCookieTextLabel(cookieTextLabel)} />
            <ButtonsWrapper className='buttons-wrapper'>
              {validateReverseButtons(reverseButtons)
                ? buttons.reverse()
                : buttons}
            </ButtonsWrapper>
          </Wrapper>
        </StickToBottom>
      </Root>
    )
  }
}
