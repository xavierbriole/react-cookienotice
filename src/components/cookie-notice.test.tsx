import { act, render } from '@testing-library/react'
import { vi } from 'vitest'

import { getCookieValue, setCookie } from '../helpers/cookies'
import CookieNotice from './cookie-notice'

vi.mock('../helpers/cookies', () => ({
  setCookie: vi.fn(),
  getCookieValue: vi.fn(),
}))

describe('CookieNotice', () => {
  beforeEach(() => {
    vi.resetAllMocks()
  })

  afterEach(() => {
    vi.useRealTimers()
  })

  describe('should render', () => {
    it('with default props', () => {
      const { asFragment } = render(
        <CookieNotice
          acceptAllButtonLabel='acceptAllButtonLabel'
          onAcceptAllButtonClick={() => {}}
          declineAllButtonLabel='declineAllButtonLabel'
          onDeclineAllButtonClick={() => {}}
          customizeButtonLabel='customizeButtonLabel'
          customizeTitleLabel='customizeTitleLabel'
          services={[
            {
              name: 'service1Name',
              description: 'service1Description',
              code: 'service1Code',
              alwaysActive: true,
            },
            {
              name: 'service2Name',
              description: 'service2Description',
              code: 'service2Code',
              alwaysActive: false,
            },
          ]}
          acceptButtonLabel='acceptButtonLabel'
          onAcceptButtonClick={() => {}}
          backButtonLabel='backButtonLabel'
          alwaysActiveLabel='alwaysActiveLabel'
          customizeAcceptAllButtonLabel='customizeAcceptAllButtonLabel'
          customizeAcceptAllTimeout={1000}
          titleLabel='titleLabel'
          descriptionLabel='descriptionLabel'
          readMoreLabel='readMoreLabel'
          readMoreLink='https://www.example.com'
          readMoreInNewTab={true}
          placement={{ vertical: 'bottom', horizontal: 'left' }}
          cookieOptions={{
            name: 'hide-notice',
            value: 'true',
            expires: 30,
            secure: false,
            httpOnly: false,
            sameSite: 'lax',
          }}
        />,
      )

      expect(asFragment()).toMatchSnapshot()
    })

    it('with no read more link', () => {
      const { asFragment } = render(
        <CookieNotice
          acceptAllButtonLabel='acceptAllButtonLabel'
          onAcceptAllButtonClick={() => {}}
          declineAllButtonLabel='declineAllButtonLabel'
          onDeclineAllButtonClick={() => {}}
          customizeButtonLabel='customizeButtonLabel'
          customizeTitleLabel='customizeTitleLabel'
          services={[
            {
              name: 'service1Name',
              description: 'service1Description',
              code: 'service1Code',
              alwaysActive: true,
            },
            {
              name: 'service2Name',
              description: 'service2Description',
              code: 'service2Code',
              alwaysActive: false,
            },
          ]}
          acceptButtonLabel='acceptButtonLabel'
          onAcceptButtonClick={() => {}}
          backButtonLabel='backButtonLabel'
          alwaysActiveLabel='alwaysActiveLabel'
          customizeAcceptAllButtonLabel='customizeAcceptAllButtonLabel'
          customizeAcceptAllTimeout={1000}
          titleLabel='titleLabel'
          descriptionLabel='descriptionLabel'
          readMoreLabel={undefined}
          readMoreLink={undefined}
          readMoreInNewTab={undefined}
          placement={{ vertical: 'bottom', horizontal: 'left' }}
          cookieOptions={{
            name: 'hide-notice',
            value: 'true',
            expires: 30,
            secure: false,
            httpOnly: false,
            sameSite: 'lax',
          }}
        />,
      )

      expect(asFragment()).toMatchSnapshot()
    })

    it.each([
      [{ vertical: 'top', horizontal: 'left' }],
      [{ vertical: 'top', horizontal: 'right' }],
      [{ vertical: 'bottom', horizontal: 'left' }],
      [{ vertical: 'bottom', horizontal: 'right' }],
    ])('with placement %s', (placement) => {
      const { asFragment } = render(
        <CookieNotice
          acceptAllButtonLabel='acceptAllButtonLabel'
          onAcceptAllButtonClick={() => {}}
          declineAllButtonLabel='declineAllButtonLabel'
          onDeclineAllButtonClick={() => {}}
          customizeButtonLabel='customizeButtonLabel'
          customizeTitleLabel='customizeTitleLabel'
          services={[
            {
              name: 'service1Name',
              description: 'service1Description',
              code: 'service1Code',
              alwaysActive: true,
            },
            {
              name: 'service2Name',
              description: 'service2Description',
              code: 'service2Code',
              alwaysActive: false,
            },
          ]}
          acceptButtonLabel='acceptButtonLabel'
          onAcceptButtonClick={() => {}}
          backButtonLabel='backButtonLabel'
          alwaysActiveLabel='alwaysActiveLabel'
          customizeAcceptAllButtonLabel='customizeAcceptAllButtonLabel'
          customizeAcceptAllTimeout={1000}
          titleLabel='titleLabel'
          descriptionLabel='descriptionLabel'
          readMoreLabel='readMoreLabel'
          readMoreLink='https://www.example.com'
          readMoreInNewTab={true}
          // @ts-ignore
          placement={placement}
          cookieOptions={{
            name: 'hide-notice',
            value: 'true',
            expires: 30,
            secure: false,
            httpOnly: false,
            sameSite: 'lax',
          }}
        />,
      )

      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('in default view', () => {
    it('should handle accept all button click', () => {
      const onAcceptAllButtonClick = vi.fn()

      const { getByText } = render(
        <CookieNotice
          acceptAllButtonLabel='acceptAllButtonLabel'
          onAcceptAllButtonClick={onAcceptAllButtonClick}
          declineAllButtonLabel='declineAllButtonLabel'
          onDeclineAllButtonClick={() => {}}
          customizeButtonLabel='customizeButtonLabel'
          customizeTitleLabel='customizeTitleLabel'
          services={[
            {
              name: 'service1Name',
              description: 'service1Description',
              code: 'service1Code',
              alwaysActive: true,
            },
            {
              name: 'service2Name',
              description: 'service2Description',
              code: 'service2Code',
              alwaysActive: false,
            },
          ]}
          acceptButtonLabel='acceptButtonLabel'
          onAcceptButtonClick={() => {}}
          backButtonLabel='backButtonLabel'
          alwaysActiveLabel='alwaysActiveLabel'
          customizeAcceptAllButtonLabel='customizeAcceptAllButtonLabel'
          customizeAcceptAllTimeout={1000}
          titleLabel='titleLabel'
          descriptionLabel='descriptionLabel'
          readMoreLabel='readMoreLabel'
          readMoreLink='https://www.example.com'
          readMoreInNewTab={true}
          placement={{ vertical: 'bottom', horizontal: 'left' }}
          cookieOptions={{
            name: 'hide-notice',
            value: 'true',
            expires: 30,
            secure: false,
            httpOnly: false,
            sameSite: 'lax',
          }}
        />,
      )

      act(() => {
        getByText('acceptAllButtonLabel').click()
      })

      expect(setCookie).toHaveBeenCalledTimes(1)
      expect(setCookie).toHaveBeenCalledWith({
        name: 'hide-notice',
        value: 'true',
        expires: 30,
        secure: false,
        httpOnly: false,
        sameSite: 'lax',
      })
      expect(onAcceptAllButtonClick).toHaveBeenCalledTimes(1)
    })

    it('should handle decline all button click', () => {
      const onDeclineAllButtonClick = vi.fn()

      const { getByText } = render(
        <CookieNotice
          acceptAllButtonLabel='acceptAllButtonLabel'
          onAcceptAllButtonClick={() => {}}
          declineAllButtonLabel='declineAllButtonLabel'
          onDeclineAllButtonClick={onDeclineAllButtonClick}
          customizeButtonLabel='customizeButtonLabel'
          customizeTitleLabel='customizeTitleLabel'
          services={[
            {
              name: 'service1Name',
              description: 'service1Description',
              code: 'service1Code',
              alwaysActive: true,
            },
            {
              name: 'service2Name',
              description: 'service2Description',
              code: 'service2Code',
              alwaysActive: false,
            },
          ]}
          acceptButtonLabel='acceptButtonLabel'
          onAcceptButtonClick={() => {}}
          backButtonLabel='backButtonLabel'
          alwaysActiveLabel='alwaysActiveLabel'
          customizeAcceptAllButtonLabel='customizeAcceptAllButtonLabel'
          customizeAcceptAllTimeout={1000}
          titleLabel='titleLabel'
          descriptionLabel='descriptionLabel'
          readMoreLabel='readMoreLabel'
          readMoreLink='https://www.example.com'
          readMoreInNewTab={true}
          placement={{ vertical: 'bottom', horizontal: 'left' }}
          cookieOptions={{
            name: 'hide-notice',
            value: 'true',
            expires: 30,
            secure: false,
            httpOnly: false,
            sameSite: 'lax',
          }}
        />,
      )

      act(() => {
        getByText('declineAllButtonLabel').click()
      })

      expect(setCookie).toHaveBeenCalledTimes(1)
      expect(setCookie).toHaveBeenCalledWith({
        name: 'hide-notice',
        value: 'true',
        expires: 30,
        secure: false,
        httpOnly: false,
        sameSite: 'lax',
      })
      expect(onDeclineAllButtonClick).toHaveBeenCalledTimes(1)
    })
  })

  describe('in customize view', () => {
    it('should handle accept button click', () => {
      const onAcceptButtonClick = vi.fn()

      const { getByText, getByLabelText } = render(
        <CookieNotice
          acceptAllButtonLabel='acceptAllButtonLabel'
          onAcceptAllButtonClick={() => {}}
          declineAllButtonLabel='declineAllButtonLabel'
          onDeclineAllButtonClick={() => {}}
          customizeButtonLabel='customizeButtonLabel'
          customizeTitleLabel='customizeTitleLabel'
          services={[
            {
              name: 'service1Name',
              description: 'service1Description',
              code: 'service1Code',
              alwaysActive: true,
            },
            {
              name: 'service2Name',
              description: 'service2Description',
              code: 'service2Code',
              alwaysActive: false,
            },
            {
              name: 'service3Name',
              description: 'service3Description',
              code: 'service3Code',
              alwaysActive: false,
            },
          ]}
          acceptButtonLabel='acceptButtonLabel'
          onAcceptButtonClick={onAcceptButtonClick}
          backButtonLabel='backButtonLabel'
          alwaysActiveLabel='alwaysActiveLabel'
          customizeAcceptAllButtonLabel='customizeAcceptAllButtonLabel'
          customizeAcceptAllTimeout={1000}
          titleLabel='titleLabel'
          descriptionLabel='descriptionLabel'
          readMoreLabel='readMoreLabel'
          readMoreLink='https://www.example.com'
          readMoreInNewTab={true}
          placement={{ vertical: 'bottom', horizontal: 'left' }}
          cookieOptions={{
            name: 'hide-notice',
            value: 'true',
            expires: 30,
            secure: false,
            httpOnly: false,
            sameSite: 'lax',
          }}
        />,
      )

      act(() => {
        getByText('customizeButtonLabel').click()
      })

      act(() => {
        getByLabelText('service2Code').click()
        getByLabelText('service3Code').click()
      })

      act(() => {
        getByText('acceptButtonLabel').click()
      })

      expect(setCookie).toHaveBeenCalledTimes(1)
      expect(setCookie).toHaveBeenCalledWith({
        name: 'hide-notice',
        value: 'true',
        expires: 30,
        secure: false,
        httpOnly: false,
        sameSite: 'lax',
      })
      expect(onAcceptButtonClick).toHaveBeenCalledTimes(1)
      expect(onAcceptButtonClick).toHaveBeenCalledWith([
        'service2Code',
        'service3Code',
        'service1Code',
      ])
    })

    it('should handle accept all button click', () => {
      vi.useFakeTimers()

      const onAcceptAllButtonClick = vi.fn()

      const { getByText } = render(
        <CookieNotice
          acceptAllButtonLabel='acceptAllButtonLabel'
          onAcceptAllButtonClick={onAcceptAllButtonClick}
          declineAllButtonLabel='declineAllButtonLabel'
          onDeclineAllButtonClick={() => {}}
          customizeButtonLabel='customizeButtonLabel'
          customizeTitleLabel='customizeTitleLabel'
          services={[
            {
              name: 'service1Name',
              description: 'service1Description',
              code: 'service1Code',
              alwaysActive: true,
            },
            {
              name: 'service2Name',
              description: 'service2Description',
              code: 'service2Code',
              alwaysActive: false,
            },
            {
              name: 'service3Name',
              description: 'service3Description',
              code: 'service3Code',
              alwaysActive: false,
            },
          ]}
          acceptButtonLabel='acceptButtonLabel'
          onAcceptButtonClick={() => {}}
          backButtonLabel='backButtonLabel'
          alwaysActiveLabel='alwaysActiveLabel'
          customizeAcceptAllButtonLabel='customizeAcceptAllButtonLabel'
          customizeAcceptAllTimeout={1000}
          titleLabel='titleLabel'
          descriptionLabel='descriptionLabel'
          readMoreLabel='readMoreLabel'
          readMoreLink='https://www.example.com'
          readMoreInNewTab={true}
          placement={{ vertical: 'bottom', horizontal: 'left' }}
          cookieOptions={{
            name: 'hide-notice',
            value: 'true',
            expires: 30,
            secure: false,
            httpOnly: false,
            sameSite: 'lax',
          }}
        />,
      )

      act(() => {
        getByText('customizeButtonLabel').click()
      })

      act(() => {
        getByText('customizeAcceptAllButtonLabel').click()
        vi.runAllTimers()
      })

      expect(setCookie).toHaveBeenCalledTimes(1)
      expect(setCookie).toHaveBeenCalledWith({
        name: 'hide-notice',
        value: 'true',
        expires: 30,
        secure: false,
        httpOnly: false,
        sameSite: 'lax',
      })
      expect(onAcceptAllButtonClick).toHaveBeenCalledTimes(1)
      expect(onAcceptAllButtonClick).toHaveBeenCalledWith([
        'service1Code',
        'service2Code',
        'service3Code',
      ])
    })
  })

  it('should get cookie on mount', () => {
    render(
      <CookieNotice
        acceptAllButtonLabel='acceptAllButtonLabel'
        onAcceptAllButtonClick={() => {}}
        declineAllButtonLabel='declineAllButtonLabel'
        onDeclineAllButtonClick={() => {}}
        customizeButtonLabel='customizeButtonLabel'
        customizeTitleLabel='customizeTitleLabel'
        services={[
          {
            name: 'service1Name',
            description: 'service1Description',
            code: 'service1Code',
            alwaysActive: true,
          },
          {
            name: 'service2Name',
            description: 'service2Description',
            code: 'service2Code',
            alwaysActive: false,
          },
        ]}
        acceptButtonLabel='acceptButtonLabel'
        onAcceptButtonClick={() => {}}
        backButtonLabel='backButtonLabel'
        alwaysActiveLabel='alwaysActiveLabel'
        customizeAcceptAllButtonLabel='customizeAcceptAllButtonLabel'
        customizeAcceptAllTimeout={1000}
        titleLabel='titleLabel'
        descriptionLabel='descriptionLabel'
        readMoreLabel='readMoreLabel'
        readMoreLink='https://www.example.com'
        readMoreInNewTab={true}
        placement={{ vertical: 'bottom', horizontal: 'left' }}
        cookieOptions={{
          name: 'hide-notice',
          value: 'true',
          expires: 30,
          secure: false,
          httpOnly: false,
          sameSite: 'lax',
        }}
      />,
    )

    expect(getCookieValue).toHaveBeenCalledTimes(1)
    expect(getCookieValue).toHaveBeenCalledWith('hide-notice')
  })

  it('should switch views', () => {
    const { getByText, getByLabelText } = render(
      <CookieNotice
        acceptAllButtonLabel='acceptAllButtonLabel'
        onAcceptAllButtonClick={() => {}}
        declineAllButtonLabel='declineAllButtonLabel'
        onDeclineAllButtonClick={() => {}}
        customizeButtonLabel='customizeButtonLabel'
        customizeTitleLabel='customizeTitleLabel'
        services={[
          {
            name: 'service1Name',
            description: 'service1Description',
            code: 'service1Code',
            alwaysActive: true,
          },
          {
            name: 'service2Name',
            description: 'service2Description',
            code: 'service2Code',
            alwaysActive: false,
          },
        ]}
        acceptButtonLabel='acceptButtonLabel'
        onAcceptButtonClick={() => {}}
        backButtonLabel='backButtonLabel'
        alwaysActiveLabel='alwaysActiveLabel'
        customizeAcceptAllButtonLabel='customizeAcceptAllButtonLabel'
        customizeAcceptAllTimeout={1000}
        titleLabel='titleLabel'
        descriptionLabel='descriptionLabel'
        readMoreLabel='readMoreLabel'
        readMoreLink='https://www.example.com'
        readMoreInNewTab={true}
        placement={{ vertical: 'bottom', horizontal: 'left' }}
        cookieOptions={{
          name: 'hide-notice',
          value: 'true',
          expires: 30,
          secure: false,
          httpOnly: false,
          sameSite: 'lax',
        }}
      />,
    )

    expect(getByText('titleLabel')).toBeInTheDocument()
    expect(getByText('descriptionLabel')).toBeInTheDocument()
    expect(getByText('readMoreLabel')).toBeInTheDocument()
    expect(getByText('acceptAllButtonLabel')).toBeInTheDocument()
    expect(getByText('customizeButtonLabel')).toBeInTheDocument()
    expect(getByText('declineAllButtonLabel')).toBeInTheDocument()

    act(() => {
      getByText('customizeButtonLabel').click()
    })

    expect(getByText('customizeTitleLabel')).toBeInTheDocument()
    expect(getByText('alwaysActiveLabel')).toBeInTheDocument()
    expect(getByLabelText('service1Code')).toBeInTheDocument()
    expect(getByLabelText('service2Code')).toBeInTheDocument()
    expect(getByText('acceptButtonLabel')).toBeInTheDocument()
    expect(getByText('customizeAcceptAllButtonLabel')).toBeInTheDocument()
    expect(getByText('backButtonLabel')).toBeInTheDocument()

    act(() => {
      getByText('backButtonLabel').click()
    })

    expect(getByText('titleLabel')).toBeInTheDocument()
    expect(getByText('descriptionLabel')).toBeInTheDocument()
    expect(getByText('readMoreLabel')).toBeInTheDocument()
    expect(getByText('acceptAllButtonLabel')).toBeInTheDocument()
    expect(getByText('customizeButtonLabel')).toBeInTheDocument()
    expect(getByText('declineAllButtonLabel')).toBeInTheDocument()
  })

  describe('should handle classNames', () => {
    it('should add hide-with-animation class', () => {
      const { getByText } = render(
        <CookieNotice
          acceptAllButtonLabel='acceptAllButtonLabel'
          onAcceptAllButtonClick={() => {}}
          declineAllButtonLabel='declineAllButtonLabel'
          onDeclineAllButtonClick={() => {}}
          customizeButtonLabel='customizeButtonLabel'
          customizeTitleLabel='customizeTitleLabel'
          services={[
            {
              name: 'service1Name',
              description: 'service1Description',
              code: 'service1Code',
              alwaysActive: true,
            },
            {
              name: 'service2Name',
              description: 'service2Description',
              code: 'service2Code',
              alwaysActive: false,
            },
          ]}
          acceptButtonLabel='acceptButtonLabel'
          onAcceptButtonClick={() => {}}
          backButtonLabel='backButtonLabel'
          alwaysActiveLabel='alwaysActiveLabel'
          customizeAcceptAllButtonLabel='customizeAcceptAllButtonLabel'
          customizeAcceptAllTimeout={1000}
          titleLabel='titleLabel'
          descriptionLabel='descriptionLabel'
          readMoreLabel='readMoreLabel'
          readMoreLink='https://www.example.com'
          readMoreInNewTab={true}
          placement={{ vertical: 'bottom', horizontal: 'left' }}
          cookieOptions={{
            name: 'hide-notice',
            value: 'true',
            expires: 30,
            secure: false,
            httpOnly: false,
            sameSite: 'lax',
          }}
        />,
      )

      const cookieNoticeBeforeAccept = document.querySelector(
        '.react-cookienotice-root',
      )

      if (!cookieNoticeBeforeAccept) {
        throw new Error('Cookie notice not found')
      }

      expect(cookieNoticeBeforeAccept.className).not.toContain(
        'hide-with-animation',
      )

      act(() => {
        getByText('acceptAllButtonLabel').click()
      })

      const cookieNoticeAfterAccept = document.querySelector(
        '.react-cookienotice-root',
      )

      if (!cookieNoticeAfterAccept) {
        throw new Error('Cookie notice not found')
      }

      expect(cookieNoticeAfterAccept.className).toContain('hide-with-animation')
    })

    it('should mount without hidden class', () => {
      // @ts-ignore
      getCookieValue.mockReturnValue('')

      render(
        <CookieNotice
          acceptAllButtonLabel='acceptAllButtonLabel'
          onAcceptAllButtonClick={() => {}}
          declineAllButtonLabel='declineAllButtonLabel'
          onDeclineAllButtonClick={() => {}}
          customizeButtonLabel='customizeButtonLabel'
          customizeTitleLabel='customizeTitleLabel'
          services={[
            {
              name: 'service1Name',
              description: 'service1Description',
              code: 'service1Code',
              alwaysActive: true,
            },
            {
              name: 'service2Name',
              description: 'service2Description',
              code: 'service2Code',
              alwaysActive: false,
            },
          ]}
          acceptButtonLabel='acceptButtonLabel'
          onAcceptButtonClick={() => {}}
          backButtonLabel='backButtonLabel'
          alwaysActiveLabel='alwaysActiveLabel'
          customizeAcceptAllButtonLabel='customizeAcceptAllButtonLabel'
          customizeAcceptAllTimeout={1000}
          titleLabel='titleLabel'
          descriptionLabel='descriptionLabel'
          readMoreLabel='readMoreLabel'
          readMoreLink='https://www.example.com'
          readMoreInNewTab={true}
          placement={{ vertical: 'bottom', horizontal: 'left' }}
          cookieOptions={{
            name: 'hide-notice',
            value: 'true',
            expires: 30,
            secure: false,
            httpOnly: false,
            sameSite: 'lax',
          }}
        />,
      )

      const cookieNotice = document.querySelector('.react-cookienotice-root')

      if (!cookieNotice) {
        throw new Error('Cookie notice not found')
      }

      expect(cookieNotice.className).not.toContain('hidden')
    })

    it('should mount with hidden class', () => {
      // @ts-ignore
      getCookieValue.mockReturnValue('true')

      render(
        <CookieNotice
          acceptAllButtonLabel='acceptAllButtonLabel'
          onAcceptAllButtonClick={() => {}}
          declineAllButtonLabel='declineAllButtonLabel'
          onDeclineAllButtonClick={() => {}}
          customizeButtonLabel='customizeButtonLabel'
          customizeTitleLabel='customizeTitleLabel'
          services={[
            {
              name: 'service1Name',
              description: 'service1Description',
              code: 'service1Code',
              alwaysActive: true,
            },
            {
              name: 'service2Name',
              description: 'service2Description',
              code: 'service2Code',
              alwaysActive: false,
            },
          ]}
          acceptButtonLabel='acceptButtonLabel'
          onAcceptButtonClick={() => {}}
          backButtonLabel='backButtonLabel'
          alwaysActiveLabel='alwaysActiveLabel'
          customizeAcceptAllButtonLabel='customizeAcceptAllButtonLabel'
          customizeAcceptAllTimeout={1000}
          titleLabel='titleLabel'
          descriptionLabel='descriptionLabel'
          readMoreLabel='readMoreLabel'
          readMoreLink='https://www.example.com'
          readMoreInNewTab={true}
          placement={{ vertical: 'bottom', horizontal: 'left' }}
          cookieOptions={{
            name: 'hide-notice',
            value: 'true',
            expires: 30,
            secure: false,
            httpOnly: false,
            sameSite: 'lax',
          }}
        />,
      )

      const cookieNotice = document.querySelector('.react-cookienotice-root')

      if (!cookieNotice) {
        throw new Error('Cookie notice not found')
      }

      expect(cookieNotice.className).toContain('hidden')
    })
  })
})
