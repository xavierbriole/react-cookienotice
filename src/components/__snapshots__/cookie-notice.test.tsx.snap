// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CookieNotice should render 1`] = `
<div
  class="react-cookienotice-root root"
>
  <div
    class="react-cookienotice-body body"
  >
    <div
      class="react-cookienotice-title title"
    >
      <span>
        titleLabel
      </span>
    </div>
    <div
      class="react-cookienotice-description description"
    >
      <span>
        descriptionLabel
      </span>
    </div>
    <a
      class="react-cookienotice-link link"
      href="https://example.com"
      rel="noreferrer"
      target="_blank"
    >
      readMoreLabel
    </a>
  </div>
  <div
    class="react-cookienotice-buttons buttons"
  >
    <button
      class="react-cookienotice-button button"
    >
      acceptButtonLabel
    </button>
    <button
      class="react-cookienotice-button button"
    >
      declineButtonLabel
    </button>
  </div>
</div>
`;
